src/Microsoft.AspNetCore.SignalR.Common/Internal/Encoders/Base64Encoder.cs

```diff
@@ -14,15 +16,22 @@ public byte[] Decode(byte[] payload)
             var buffer = new ReadOnlyMemory<byte>(payload);
             LengthPrefixedTextMessageParser.TryParseMessage(ref buffer, out var message);
 
-            return Convert.FromBase64String(Encoding.UTF8.GetString(message.ToArray()));
+            Span<byte> decoded = new byte[Base64.GetMaxDecodedFromUtf8Length(message.Length)];
+            var status = Base64.DecodeFromUtf8(message.Span, decoded, out _, out var written);
+            Debug.Assert(status == OperationStatus.Done);
+
+            return decoded.Slice(0, written).ToArray();
```

Comment:
Is it necessary to copy the span (after decoded) into an array? What if the payload was `Span<byte>` instead of byte[] and the function returned a `Span<byte>`?

It looks like the only place Decode is called is here:
https://github.com/aspnet/SignalR/blob/dev/src/Microsoft.AspNetCore.SignalR.Common/Internal/HubProtocolReaderWriter.cs#L24

You could change the TryParseMessage to take in a ReadOnlySpan<byte> instead of ReadOnlyMemory<byte>.
i.e. Change [this](https://github.com/aspnet/SignalR/blob/dev/src/Microsoft.AspNetCore.SignalR.Common/Internal/Protocol/IHubProtocol.cs#L16) to:
```C#
bool TryParseMessages(ReadOnlySpan<byte> input, IInvocationBinder binder, out IList<HubMessage> messages);
```

It looks like you get the span out of it anyway:
https://github.com/aspnet/SignalR/blob/dev/src/Microsoft.AspNetCore.SignalR.Common/Internal/Formatters/TextMessageParser.cs#L14

Let me know if that would work or if I am missing something in how IHubProtocol and the TryParseMessage(s) methods are used.

cc @KrzysztofCwalina

Reply:
I did not want this change to explode. I am trying it now though.

Reply:
I agree with @ahsonkhan. The whole point of these APIs is to avoid allocating and copying data (as it happens in the call to ToArray).